drivers/pcie/endpoint/pcie_ep_common.c

```diff
@@ -90,21 +142,149 @@ int pcie_ep_xfer_data_memcpy(const struct device *dev, uint64_t pcie_addr,
 		return mapped_size;
 	}
 
+	/*
+	 * In second attempt, we must have mapped entire size,
+	 * if not just quit here before attempting memcpy
+	 */
+	if (mapped_size != unmapped_size) {
+		pcie_ep_unmap_addr(dev, mapped_addr);
+		return -EIO;
+	}
+
 	ret = pcie_ep_mapped_copy(mapped_addr,
 				  ((uintptr_t)local_addr) + xfer_size,
 				  mapped_size, dir);
 
+	/* Flush the PCIe writes upon successful memcpy */
+	if (!ret && (dir == DEVICE_TO_HOST)) {
+		sys_read8(mapped_addr);
+	}
+
 	pcie_ep_unmap_addr(dev, mapped_addr);
 
 	/* Check if mapped_copy succeeded */
 	if (ret < 0) {
 		return ret;
 	}
 
-	/* In second attempt, we must have completed data transfer */
+	return 0;
+}
+
+/*
+ * Helper API to achieve data transfer with DMA operation through
+ * PCIe outbound memory, this API is based off pcie_ep_xfer_data_memcpy,
+ * here we use "system dma" instead of memcpy
+ */
+int pcie_ep_xfer_data_dma(const struct device *dev, uint64_t pcie_addr,
+			  uintptr_t *local_addr, uint32_t size,
+			  enum pcie_ob_mem_type ob_mem_type,
+			  enum xfer_direction dir)
+{
+	uint64_t mapped_addr;
+	int mapped_size, ret;
+	uint32_t xfer_size, unmapped_size;
+	uint32_t dummy_data;	/* For explicit dummy PCIe read */
+
+	mapped_size = pcie_ep_map_addr(dev, pcie_addr, &mapped_addr,
+				       size, ob_mem_type);
+
+	/* Check if outbound memory mapping succeeded */
+	if (mapped_size < 0) {
+		return mapped_size;
+	}
+
+	ret = pcie_ep_dma_xfer(dev, mapped_addr, (uintptr_t)local_addr,
+			       mapped_size, dir);
+
+	/* Flush the PCIe writes upon successful DMA */
+	if (!ret && (dir == DEVICE_TO_HOST)) {
+		if (IS_ENABLED(CONFIG_64BIT) || !(mapped_addr >> 32)) {
+			sys_read8(mapped_addr);
+		}
+	}
+
+	pcie_ep_unmap_addr(dev, mapped_addr);
+
+	/* Check if dma succeeded */
```

Comment:
I would check this first. Actually, I would revise how you treat error cases here when you have a valid mapped_addr:

```
if (ret < 0) {
        goto out;
}
```

...

at the end:
```
out:
        pcie_ep_unmap_addr(dev, mapped_addr);
        return ret;
```

this will simplify the code as you'll call pcie_ep_unmap_addr() only once in the whole function.
and lines 284 to 286 could be removed as well.

Actually, pcie_ep_mapped_copy() could follow the same error management logic.

Reply:
I see. I will check and update.

Reply:
The requirement is to unmap the `pcie_addr` once it is mapped, even in case of data transfer failure. so, it is not done in failure path for data transfer; rather it is done unconditionally once mapping is successful (and data transfer is done irrespective of result). So, suggested goto design does not fit in.

Reply:
check again, that can be refactored.

Reply:
@tbursztyka I agree that code can be refactored, but the comment - "you'll call pcie_ep_unmap_addr() only once in the whole function" - does not hold true. To explain more, there are 2 cases:
1) In case of split transfer (look for "map remaining size and complete remaining data transfer" in comments for both `pcie_ep_xfer_data_memcpy` and `pcie_ep_xfer_data_dma`), we must unmap before attempting second half of the transfer.
2) In case of explicit PCIe read in `pcie_ep_xfer_data_dma`, before the read with `pcie_ep_xfer_data_memcpy` , we must unmap.
I will refactor the code for the cases where we have valid `mapped_addr` per the suggestion and submit.

Reply:
@tbursztyka updated error management logic per the suggestion.